ipinfo_django/middleware.py

```diff
@@ -32,8 +34,39 @@ def process_request(self, request):
             LOGGER.error(traceback.format_exc())
 
     def is_bot(self, request):
-        """Whether or not the request user-agent self-identifies as a bot"""
-        lowercase_user_agent = request.META.get("HTTP_USER_AGENT", "").lower()
-        return (
-            "bot" in lowercase_user_agent or "spider" in lowercase_user_agent
-        )
+        return is_bot(request)
+
+
+class IPinfoAsyncMiddleware:
+    sync_capable = False
+    async_capable = True
+
+    def __init__(self, get_response):
+        """Initialize class, get settings, and create the cache."""
+        self.get_response = get_response
+
+        self.filter = getattr(settings, "IPINFO_FILTER", self.is_bot)
+
+        ipinfo_token = getattr(settings, "IPINFO_TOKEN", None)
+        ipinfo_settings = getattr(settings, "IPINFO_SETTINGS", {})
+        self.ipinfo = ipinfo.getHandlerAsync(ipinfo_token, **ipinfo_settings)
+
+    def __call__(self, request):
+        return self.__acall__(request)
+
+    async def __acall__(self, request):
+        """Middleware hook that acts on and modifies request object."""
+        try:
+            if self.filter and self.filter(request):
+                request.ipinfo = None
+            else:
+                request.ipinfo = await self.ipinfo.getDetails(get_ip(request))
```

Comment:
Could we also have the sync handler do this `get_ip` part?

Reply:
Do you mean here https://github.com/valentinogagliardi/django/blob/feature/async-middleware/ipinfo_django/middleware.py#L31?

Reply:
Yep!